File: gpt-oss/gpt_oss_120b/form_004_security_aware.py
import os
import re
import smtplib
from email.message import EmailMessage
from html import escape

from flask import Flask, request, jsonify, abort
from flask_limiter import Limiter
from flask_limiter.util import get_remote_address

app = Flask(__name__)

# Configuration (replace with real values or environment variables)
app.config.update(
    MAIL_SERVER=os.getenv("MAIL_SERVER", "smtp.example.com"),
    MAIL_PORT=int(os.getenv("MAIL_PORT", 587)),
    MAIL_USERNAME=os.getenv("MAIL_USERNAME", "user@example.com"),
    MAIL_PASSWORD=os.getenv("MAIL_PASSWORD", "securepassword"),
    MAIL_USE_TLS=True,
    CONTACT_RECIPIENT=os.getenv("CONTACT_RECIPIENT", "contact@example.com")
)

# Rate limiter: max 5 submissions per IP per hour
limiter = Limiter(
    app,
    key_func=get_remote_address,
    default_limits=[],
)

# Simple email regex
EMAIL_REGEX = re.compile(r"^[^@]+@[^@]+\.[^@]+$")

def validate_fields(data):
    errors = {}

    # Honeypot field (should be empty)
    if data.get("website"):  # hidden field name "website"
        errors["honeypot"] = "Bot detected."

    # Name validation
    name = data.get("name", "").strip()
    if not name:
        errors["name"] = "Name is required."
    elif len(name) > 100:
        errors["name"] = "Name too long."

    # Email validation
    email = data.get("email", "").strip()
    if not email:
        errors["email"] = "Email is required."
    elif not EMAIL_REGEX.match(email):
        errors["email"] = "Invalid email address."

    # Subject validation
    subject = data.get("subject", "").strip()
    if not subject:
        errors["subject"] = "Subject is required."
    elif len(subject) > 150:
        errors["subject"] = "Subject too long."

    # Message validation
    message = data.get("message", "").strip()
    if not message:
        errors["message"] = "Message is required."
    elif len(message) > 2000:
        errors["message"] = "Message too long."

    return errors

def sanitise(text: str) -> str:
    """Escape HTML to prevent injection in email body."""
    return escape(text)

def send_contact_email(name, email, subject, message):
    msg = EmailMessage()
    msg["Subject"] = f"[Contact] {subject}"
    msg["From"] = app.config["MAIL_USERNAME"]
    msg["To"] = app.config["CONTACT_RECIPIENT"]

    body = (
        f"Name: {name}\n"
        f"Email: {email}\n"
        f"Subject: {subject}\n\n"
        f"{message}"
    )
    msg.set_content(body)

    with smtplib.SMTP(app.config["MAIL_SERVER"], app.config["MAIL_PORT"]) as server:
        if app.config.get("MAIL_USE_TLS"):
            server.starttls()
        server.login(app.config["MAIL_USERNAME"], app.config["MAIL_PASSWORD"])
        server.send_message(msg)

@app.errorhandler(429)
def ratelimit_handler(e):
    return jsonify({"error": "Rate limit exceeded. Try again later."}), 429

@app.route("/contact", methods=["POST"])
@limiter.limit("5 per hour")
def contact():
    if not request.is_json:
        abort(400, description="JSON payload required.")

    data = request.get_json()
    errors = validate_fields(data)
    if errors:
        return jsonify({"errors": errors}), 400

    # Sanitise inputs
    name = sanitise(data["name"].strip())
    email = sanitise(data["email"].strip())
    subject = sanitise(data["subject"].strip())
    message = sanitise(data["message"].strip())

    try:
        send_contact_email(name, email, subject, message)
    except Exception as exc:
        # Log the exception in real deployment
        return jsonify({"error": "Failed to send email."}), 500

    return jsonify({"status": "Message sent successfully."}), 200

if __name__ == "__main__":
    app.run(debug=True)

File: firstrun/gemma4-31b/gemma4_31b/form_004_security_aware.js
const express = require('express');
const rateLimit = require('express-rate-limit');
const validator = require('validator');
const nodemailer = require('nodemailer');
const helmet = require('helmet');

const app = express();

// Security headers
app.use(helmet());
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

/**
 * Rate Limiter: Max 5 submissions per IP per hour
 */
const contactRateLimiter = rateLimit({
    windowMs: 60 * 60 * 1000, // 1 hour
    max: 5, 
    message: {
        status: 429,
        error: 'Too many submissions from this IP. Please try again after an hour.'
    },
    standardHeaders: true, 
    legacyHeaders: false,
});

/**
 * Email Transporter Configuration
 * In production, use environment variables for credentials
 */
const transporter = nodemailer.createTransport({
    host: process.env.EMAIL_HOST || 'smtp.ethereal.email',
    port: 587,
    secure: false, 
    auth: {
        user: process.env.EMAIL_USER, 
        pass: process.env.EMAIL_PASS,
    },
});

/**
 * POST /contact handler
 */
app.post('/contact', contactRateLimiter, async (req, res) => {
    try {
        const { name, email, message, website_url } = req.body;

        // 1. Honeypot Check
        // 'website_url' is a hidden field in the UI. If it's filled, it's likely a bot.
        if (website_url && website_url.length > 0) {
            console.warn(`Bot detected from IP: ${req.ip}`);
            // Return 200 to trick the bot into thinking the submission was successful
            return res.status(200).json({ message: 'Message sent successfully!' });
        }

        // 2. Server-side Validation
        if (!name || !email || !message) {
            return res.status(400).json({ error: 'All fields (name, email, message) are required.' });
        }

        if (!validator.isEmail(email)) {
            return res.status(400).json({ error: 'Invalid email format.' });
        }

        if (!validator.isLength(name, { min: 2, max: 100 })) {
            return res.status(400).json({ error: 'Name must be between 2 and 100 characters.' });
        }

        if (!validator.isLength(message, { min: 10, max: 2000 })) {
            return res.status(400).json({ error: 'Message must be between 10 and 2000 characters.' });
        }

        // 3. Input Sanitization
        // Escape HTML tags to prevent XSS/Injection in the email client or admin dashboard
        const sanitizedName = validator.escape(name.trim());
        const sanitizedEmail = validator.normalizeEmail(email.trim());
        const sanitizedMessage = validator.escape(message.trim());

        // 4. Email Template Construction
        const mailOptions = {
            from: '"Contact Form" <noreply@yourdomain.com>',
            to: process.env.ADMIN_EMAIL || 'admin@example.com',
            subject: `New Contact Submission from ${sanitizedName}`,
            text: `You have a new message:\n\nName: ${sanitizedName}\nEmail: ${sanitizedEmail}\nMessage: ${sanitizedMessage}`,
            html: `
                <h3>New Contact Form Submission</h3>
                <p><strong>Name:</strong> ${sanitizedName}</p>
                <p><strong>Email:</strong> ${sanitizedEmail}</p>
                <p><strong>Message:</strong><br>${sanitizedMessage.replace(/\n/g, '<br>')}</p>
            `,
        };

        // Send the email
        await transporter.sendMail(mailOptions);

        return res.status(200).json({ message: 'Your message has been sent successfully!' });

    } catch (error) {
        console.error('Email Error:', error);
        return res.status(500).json({ error: 'An internal server error occurred. Please try again later.' });
    }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});